server/src/test/java/org/elasticsearch/search/internal/ContextIndexSearcherTests.java

```diff
@@ -651,34 +518,365 @@ public boolean isCacheable(LeafReaderContext ctx) {
                             }
                         };
                     }
+                }, manager);
+                assertTrue(contextIndexSearcher.timeExceeded());
+                assertThat(called[0], equalTo(true));
+            } finally {
+                if (executor != null) {
+                    terminate(executor);
+                }
+            }
+        }
+    }
 
-                    @Override
-                    public String toString(String field) {
-                        return "query";
+    /**
+     * Simulate one or more exceptions being thrown while collecting, through a custom query that throws IOException in its Weight#scorer.
+     * Verify that the slices that had to wait because there were no available threads in the pool are not started following the exception,
+     * which triggers a cancellation of all the tasks that are part of the running search.
+     * Simulate having N threads busy doing other work (e.g. other searches) otherwise all slices can be executed directly, given that
+     * the number of slices is dependent on the max pool size.
+     */
+    public void testCancelSliceTasksOnException() throws Exception {
```

Comment:
I have adjusted and simplified some existing tests, as well as added two tests are cancellation of tasks on exception / timeout. I wonder if there's more tests that we'd like to add to verify the behaviour of the new thread pool in combination with context index searcher.